helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -398,21 +342,71 @@ private PartitionStateMap computeIntermediatePartitionState(ResourceControllerDa
     // less than the threshold. Otherwise, only allow downward-transition load balance
     boolean onlyDownwardLoadBalance = partitionCount > threshold;
 
-    loadbalanceThrottledPartitions = loadRebalance(resource, currentStateOutput,
-        bestPossiblePartitionStateMap, throttleController, intermediatePartitionStateMap,
-        partitionsNeedLoadBalance, currentStateOutput.getCurrentStateMap(resourceName),
-        onlyDownwardLoadBalance, stateModelDef, cache);
+    chargePendingTransition(resource, currentStateOutput, throttleController, cache, preferenceLists, stateModelDef, intermediatePartitionStateMap);
+
+    // Sort partitions in case of urgent partition need to take the quota first.
+    List<Partition> partitions = new ArrayList<>(resource.getPartitions());
+    Collections.sort(partitions, new PartitionPriorityComparator(bestPossiblePartitionStateMap.getStateMap(),
+        currentStateOutput.getCurrentStateMap(resourceName), stateModelDef.getTopState()));
+    for (Partition partition : partitions) {
+      Map<String, String> derivedCurrentStateMap = currentStateOutput.getCurrentStateMap(resourceName, partition)
+          .entrySet()
+          .stream()
+          .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
+      List<String> preferenceList = preferenceLists.get(partition.getPartitionName());
+      Map<String, Integer> requiredState = getRequiredStates(resourceName, cache, preferenceList);
+      List<Message> messagesToThrottle = new ArrayList<>(resourceMessageMap.get(partition));
+      if (messagesToThrottle == null || messagesToThrottle.isEmpty()) {
+        continue;
+      }
+      Collections.sort(messagesToThrottle,
+          new MessagePriorityComparator(preferenceList, stateModelDef.getStatePriorityMap()));
+      for (Message message : messagesToThrottle) {
+        RebalanceType rebalanceType = getRebalanceTypePerMessage(requiredState, message, derivedCurrentStateMap);
+
+        // Number of states required by StateModelDefinition are not satisfied, need recovery
```

Comment:
This comment needs to be right above line 365 where the rebalanceType is determined.